### PR TITLE
[contract-client] refactor

### PR DESCRIPTION
**Pull request by niklasad1**
Hi,
This is a step toward closing [#7821](https://github.com/paritytech/parity/issues/7821) but after realizing that it requires major refactoring in ```ethcore``` we decided to skip that because it also dependent on [#7038](https://github.com/paritytech/parity/pull/7038).

Thus, the PR provides the following:
* A new crate ```contract-client``` that is meant handle interactions to the registrar contract instead of having multiple interfaces and moves this functionality from ```hash-fetch```

### CODE REVIEW

File: contract_client/src/contract_client.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
+use futures::{Future, future};
+use ethabi::{Address, Bytes};
+use std::sync::Arc;
+use keccak_hash::keccak;
+
+use_contract!(registry, "Registry", "res/registrar.json");
+
+// This need to be documented why it is concatenated with the hash
+const MAGIC_CONST: &'static str = "A";
```
Comment (rphmeier) [inline contract_client/src/contract_client.rs]: I think it's meant to be an "A" record as in DNS

**Comment by rphmeier** [comment]
> 1. synchronous (to be used by LightClient)
> 2. asyncronous (to be used by everyhing else)

This seems backwards. The light client by nature serves contract call requests asynchronously (as it has to go to the network), while full nodes don't even have a futures-based API for doing so.

File: contract_client/src/contract_client.rs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+		}
+		.map_err(|e| e.to_string()));
+
+		Box::new(future)
+	}
+}
+
+/// Syncronous contract interface.
+/// Should execute transaction using current blockchain state.
+pub trait ContractClient: Send + Sync {
```
[the remark this answers is not in the document]
Comment (rphmeier) [reply contract_client/src/contract_client.rs]: and is this trait actually used anywhere?

**Comment by rphmeier** [comment]
`contract-client` as a crate name is way too general. Maybe `registrar-wrapper` or something which describes its intended use.

File: contract_client/src/contract_client.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (niklasad1) [reply contract_client/src/contract_client.rs]: No, it is not used anywhere but ideally this should be used ```ethcore``` which is synchronous if I'm not mistaken!

**Comment by niklasad1** [comment]
> This seems backwards. The light client by nature serves contract call requests asynchronously (as it has to go to the network), while full nodes don't even have a futures-based API for doing so.

Ok, maybe I misunderstood Marek then but at least the ```API``` that ```ethcore/miner``` is using is synchronous by only looking at the [method signatures](https://github.com/paritytech/parity/blob/master/ethcore/src/client/traits.rs#L274-#L283). Maybe that requires another trait for later but that might change depending on the other PR if I understand correct.

**Comment by niklasad1** [comment]
> contract-client as a crate name is way too general and not a good description of its contents. Maybe registrar-wrapper or something which describes its intended use.

I agree and disagree :smile: as long as trait contains the ```ContractClient``` I want something in the name to indicate that. However if it's possible to remove ```ContractClient``` I'd be happy to rename it registrar-wrapper / registrar-client or something similar.

**Comment by 0x7CFE** [comment]
@niklasad1, #7038 is almost complete and needs final review before merge. I hope I would not block you for a long time.

**Comment by rphmeier** [comment]
@niklasad1 or even rename the `ContractClient` trait to something else. It's a terrible name for the trait and the crate.

**Comment by niklasad1** [comment]
@rphmeier 
What do you think about re-naming the crate to ```registrar``` and re-naming the ```trait RegistrarClient```?

Because the crate should encapsulate the registrar with a uniform interface for client and it makes sense? Yepp, naming is one of the hardest problems in CS :zap:

**Comment by rphmeier** [comment]
@niklasad1 sounds fine to me

**Comment by niklasad1** [comment]
@tomusdrw 

> Synchronous variant is not really used anywhere Is that expected?

Yes, that is correct. The intention was to use `Synchronous` in `èthcore/miner` because it has its own interface to the `registrar` with different types. But I will look into it now when [#7038](https://github.com/paritytech/parity/pull/7038) is merged.

> Also in some places we could use more concrete types than Box<Future...>, but that's a minor thing, probably an improvement we can do in future.

Ok, do you want me create an issue for that?

**Comment by tomusdrw** [comment]
@niklasad1 don't bother, it's not a hot code anyway, just good that it will be possible in future.

File: registrar/src/registrar.rs [hunk 3]
```diff
@@ -0,0 +1,61 @@
+use futures::{Future, future, IntoFuture};
```
Comment (andresilva) [inline registrar/src/registrar.rs]: Missing license header.

File: registrar/src/registrar.rs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+use ethabi::{Address, Bytes};
+use std::sync::Arc;
+use keccak_hash::keccak;
+
+use_contract!(registry, "Registry", "res/registrar.json");
+
+// Maps a domain name to IPv4 address
```
Comment (andresilva) [inline registrar/src/registrar.rs]: `Maps a domain name to an Ethereum address`

**Pull request merged**